
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="..">
      
      
        <link rel="next" href="../metrics_reference/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.14">
    
    
      
        <title>Tutorials for LZGraphs - LZGraph Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.10ba22f1.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#tutorials-for-lzgraphs" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="LZGraph Documentation" class="md-header__button md-logo" aria-label="LZGraph Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            LZGraph Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Tutorials for LZGraphs
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="LZGraph Documentation" class="md-nav__button md-logo" aria-label="LZGraph Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    LZGraph Documentation
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Welcome to LZGraphs Documentation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Tutorials for LZGraphs
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Tutorials for LZGraphs
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#getting-started-with-graph-construction" class="md-nav__link">
    <span class="md-ellipsis">
      Getting Started with Graph Construction
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Getting Started with Graph Construction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#amino-acid-positional-and-nucleotide-double-positional-graphs" class="md-nav__link">
    <span class="md-ellipsis">
      Amino Acid Positional and Nucleotide Double Positional Graphs
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Amino Acid Positional and Nucleotide Double Positional Graphs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#essential-points" class="md-nav__link">
    <span class="md-ellipsis">
      Essential Points:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#constructing-a-naive-lzgraph" class="md-nav__link">
    <span class="md-ellipsis">
      Constructing a Naive LZGraph
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exploring-graph-attributes" class="md-nav__link">
    <span class="md-ellipsis">
      Exploring Graph Attributes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exploring Graph Attributes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#amino-acid-positional-and-nucleotide-double-positional-graphs_1" class="md-nav__link">
    <span class="md-ellipsis">
      Amino Acid Positional and Nucleotide Double Positional Graphs
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Amino Acid Positional and Nucleotide Double Positional Graphs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#exploring-graph-nodes-and-edges" class="md-nav__link">
    <span class="md-ellipsis">
      Exploring Graph Nodes and Edges
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#distribution-of-observed-sequence-lengths" class="md-nav__link">
    <span class="md-ellipsis">
      Distribution of Observed Sequence Lengths:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exploring-graph-initial-and-terminal-states" class="md-nav__link">
    <span class="md-ellipsis">
      Exploring Graph Initial and Terminal States:
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exploring Graph Initial and Terminal States:">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#getting-all-the-initial-states" class="md-nav__link">
    <span class="md-ellipsis">
      Getting all the initial states
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#getting-all-the-terminal-states" class="md-nav__link">
    <span class="md-ellipsis">
      Getting all the terminal states
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#graph-v-and-j-marginal-distributions" class="md-nav__link">
    <span class="md-ellipsis">
      Graph V and J Marginal Distributions:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#calculating-the-lzcentrality" class="md-nav__link">
    <span class="md-ellipsis">
      Calculating the LZCentrality
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#calculating-the-k1000-diversity-index" class="md-nav__link">
    <span class="md-ellipsis">
      Calculating the K1000 Diversity Index
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#deriving-sequence-generation-probability-pgen" class="md-nav__link">
    <span class="md-ellipsis">
      Deriving Sequence Generation Probability (Pgen)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generating-new-sequences" class="md-nav__link">
    <span class="md-ellipsis">
      Generating New Sequences
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#feature-vector-from-graph-eigen-centrality" class="md-nav__link">
    <span class="md-ellipsis">
      Feature Vector From Graph (Eigen Centrality)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plotting-graph-related-features" class="md-nav__link">
    <span class="md-ellipsis">
      Plotting Graph Related Features
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plotting Graph Related Features">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ancestors-descendants-curves-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Ancestors Descendants Curves Plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence-possible-paths-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Sequence Possible Paths Plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#node-genomic-variability-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Node Genomic Variability Plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#edge-genomic-variability-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Edge Genomic Variability Plot
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bow-vectorizer" class="md-nav__link">
    <span class="md-ellipsis">
      BOW Vectorizer
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../metrics_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LZ Metrics
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../misc_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Misc Functions
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualize_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visualize Functions
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../utilities_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Utilities Functions
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../nodedgesaturation_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Node Edge Saturation Functions
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../bowencoder_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    BOW Encoder Functions
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../lzgraphbase_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LZGraph Base Class
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../ndplgraph_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    NDPLZGraph Class
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../aaplzgraph_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    AAPLZGraph Class
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#getting-started-with-graph-construction" class="md-nav__link">
    <span class="md-ellipsis">
      Getting Started with Graph Construction
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Getting Started with Graph Construction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#amino-acid-positional-and-nucleotide-double-positional-graphs" class="md-nav__link">
    <span class="md-ellipsis">
      Amino Acid Positional and Nucleotide Double Positional Graphs
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Amino Acid Positional and Nucleotide Double Positional Graphs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#essential-points" class="md-nav__link">
    <span class="md-ellipsis">
      Essential Points:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#constructing-a-naive-lzgraph" class="md-nav__link">
    <span class="md-ellipsis">
      Constructing a Naive LZGraph
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exploring-graph-attributes" class="md-nav__link">
    <span class="md-ellipsis">
      Exploring Graph Attributes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exploring Graph Attributes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#amino-acid-positional-and-nucleotide-double-positional-graphs_1" class="md-nav__link">
    <span class="md-ellipsis">
      Amino Acid Positional and Nucleotide Double Positional Graphs
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Amino Acid Positional and Nucleotide Double Positional Graphs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#exploring-graph-nodes-and-edges" class="md-nav__link">
    <span class="md-ellipsis">
      Exploring Graph Nodes and Edges
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#distribution-of-observed-sequence-lengths" class="md-nav__link">
    <span class="md-ellipsis">
      Distribution of Observed Sequence Lengths:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exploring-graph-initial-and-terminal-states" class="md-nav__link">
    <span class="md-ellipsis">
      Exploring Graph Initial and Terminal States:
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exploring Graph Initial and Terminal States:">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#getting-all-the-initial-states" class="md-nav__link">
    <span class="md-ellipsis">
      Getting all the initial states
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#getting-all-the-terminal-states" class="md-nav__link">
    <span class="md-ellipsis">
      Getting all the terminal states
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#graph-v-and-j-marginal-distributions" class="md-nav__link">
    <span class="md-ellipsis">
      Graph V and J Marginal Distributions:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#calculating-the-lzcentrality" class="md-nav__link">
    <span class="md-ellipsis">
      Calculating the LZCentrality
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#calculating-the-k1000-diversity-index" class="md-nav__link">
    <span class="md-ellipsis">
      Calculating the K1000 Diversity Index
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#deriving-sequence-generation-probability-pgen" class="md-nav__link">
    <span class="md-ellipsis">
      Deriving Sequence Generation Probability (Pgen)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generating-new-sequences" class="md-nav__link">
    <span class="md-ellipsis">
      Generating New Sequences
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#feature-vector-from-graph-eigen-centrality" class="md-nav__link">
    <span class="md-ellipsis">
      Feature Vector From Graph (Eigen Centrality)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plotting-graph-related-features" class="md-nav__link">
    <span class="md-ellipsis">
      Plotting Graph Related Features
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plotting Graph Related Features">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ancestors-descendants-curves-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Ancestors Descendants Curves Plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence-possible-paths-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Sequence Possible Paths Plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#node-genomic-variability-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Node Genomic Variability Plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#edge-genomic-variability-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Edge Genomic Variability Plot
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bow-vectorizer" class="md-nav__link">
    <span class="md-ellipsis">
      BOW Vectorizer
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="tutorials-for-lzgraphs">Tutorials for LZGraphs</h1>
<p>Welcome to the tutorials page for LZGraphs, where you will find step-by-step guides on leveraging the power of LZGraphs for T-cell receptor beta chain (TCRB) repertoire analysis. Whether you are constructing graphs or exploring advanced features, these tutorials are designed to enhance your understanding and utilization of LZGraphs in your research endeavors.</p>
<h2 id="getting-started-with-graph-construction">Getting Started with Graph Construction</h2>
<p>LZGraphs offers various graph types to suit your analysis needs, including Amino Acid Positional, Nucleotide Double Positional, and Naive graphs. Each graph type is tailored to specific aspects of TCRB repertoire analysis, providing a versatile toolkit for your research.</p>
<h3 id="amino-acid-positional-and-nucleotide-double-positional-graphs">Amino Acid Positional and Nucleotide Double Positional Graphs</h3>
<p>These graph types are pivotal for in-depth TCRB repertoire analysis. Follow the example below to construct an Amino Acid Positional graph. The process is analogous for Nucleotide Double Positional Graphs, with minor adjustments as noted.</p>
<h4 id="essential-points">Essential Points:</h4>
<ul>
<li>For Nucleotide Double Positional Graphs, use "cdr3_rearrangement" as the column name for sequences.</li>
<li>For Amino Acid Positional Graphs, the sequence column should be "cdr3_amino_acid".</li>
<li>V and J genes/alleles are categorized under "V" and "J" columns, respectively. This genetic information enriches the graph and enhances sequence generation and genomic insights.</li>
<li>Including V/J genes/alleles is optional but recommended for comprehensive analysis.</li>
</ul>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="kn">from</span> <span class="nn">LZGraphs</span> <span class="kn">import</span> <span class="n">AAPLZGraph</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="c1"># Load your data</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;/path/to/data/file.csv&quot;</span><span class="p">)</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="c1"># Construct the graph</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="n">lzgraph</span> <span class="o">=</span> <span class="n">AAPLZGraph</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
Verbose:
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>Gene Information Loaded.. |  0.01  Seconds
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>Graph Constructed.. |  0.94  Seconds
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>Graph Metadata Derived.. |  0.94  Seconds
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>Individual Subpattern Empirical Probability Derived.. |  0.98  Seconds
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>Graph Edge Weight Normalized.. |  1.0  Seconds
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>Graph Edge Gene Weights Normalized.. |  1.13  Seconds
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a>Terminal State Map Derived.. |  1.2  Seconds
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a>Individual Subpattern Empirical Probability Derived.. |  1.37  Seconds
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a>Terminal State Map Derived.. |  1.37  Seconds
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a><span class="gh">LZGraph Created Successfully.. |  1.37  Seconds</span>
<a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a><span class="gh">==============================</span>
</code></pre></div>
The data file is a pandas DataFrame that includes the
sequences under a column named "cdr3_amino_acid".
Optionally, it can also have "V" and "J" columns containing 
relevant genomic information such as Gene/Family/Allele.
Any one of these columns can be used to create a graph, 
with functionality constrained by the genetic information on the edges.</p>
<p>Example of the input DataFrame:</p>
<table>
<thead>
<tr>
<th>cdr3_amino_acid</th>
<th>V</th>
<th>J</th>
</tr>
</thead>
<tbody>
<tr>
<td>CASSLEPSGGTDTQYF</td>
<td>TRBV16-1*01</td>
<td>TRBJ1-2*01</td>
</tr>
<tr>
<td>CASSDTSGGTDTQYF</td>
<td>TRBV1-1*01</td>
<td>TRBJ1-5*01</td>
</tr>
<tr>
<td>CASSLEPQTFTDTFFF</td>
<td>TRBV16-1*01</td>
<td>TRBJ2-7*01</td>
</tr>
</tbody>
</table>
<h3 id="constructing-a-naive-lzgraph">Constructing a Naive LZGraph</h3>
<p>Below is an example of constructing a Naive LZGraph.</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="kn">from</span> <span class="nn">LZGraphs.Graphs.Naive</span> <span class="kn">import</span> <span class="n">NaiveLZGraph</span><span class="p">,</span> <span class="n">generate_dictionary</span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="n">list_of_sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;TGTGCCAGCAGGGCGGGATACGCAGTATTT&#39;</span><span class="p">,</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>                     <span class="s1">&#39;TGTGCCAGCAGCCAGCATCGTCGCAGTATTTT&#39;</span>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>                     <span class="s1">&#39;TGTGCCAGCAGCCAGCAGGGCCGGGATACGCAGTATTTT&#39;</span>
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>                     <span class="o">...</span><span class="p">]</span>
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a><span class="c1"># generate all possible nodes with kmers up to length 6                 </span>
<a id="__codelineno-2-9" name="__codelineno-2-9" href="#__codelineno-2-9"></a><span class="n">dictionary</span> <span class="o">=</span> <span class="n">generate_dictionary</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span>
<a id="__codelineno-2-10" name="__codelineno-2-10" href="#__codelineno-2-10"></a><span class="c1"># construct a NaiveLZGraph from your list_of_sequences based on the generated dictionary</span>
<a id="__codelineno-2-11" name="__codelineno-2-11" href="#__codelineno-2-11"></a><span class="n">lzgraph</span> <span class="o">=</span> <span class="n">NaiveLZGraph</span><span class="p">(</span><span class="n">list_of_sequences</span><span class="p">,</span> <span class="n">dictionary</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
Note:</p>
<ul>
<li>The <code>generate_dictionary</code> function takes an integer <code>(X)</code> and generates all possible patterns that can be constructed from <code>(X)</code> nucleotides. For CDR3 sequences, we suggest using <code>X=6</code> as the maximum observed sub-pattern length in the CDR3 region is 6.</li>
<li>The dictionary can be replaced with any number of sub-patterns if not all combinations are desired.</li>
<li>The Naive LZGraph is most relevant in the context of nucleotides, as amino acid CDR3 sequences result in sparse graphs.</li>
</ul>
<hr />
<h2 id="exploring-graph-attributes">Exploring Graph Attributes</h2>
<p>After constructing an LZGraph, there are several attributes available to quickly infer repertoire statistical properties as well as explore the graph object itself, built above the networkx DiGraph class. Below are examples of how you can evaluate and explore those attributes.</p>
<h3 id="amino-acid-positional-and-nucleotide-double-positional-graphs_1">Amino Acid Positional and Nucleotide Double Positional Graphs</h3>
<h4 id="exploring-graph-nodes-and-edges">Exploring Graph Nodes and Edges</h4>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="c1"># Retrieving the list of all graph nodes / edges</span>
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="n">node_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">lzgraph</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="n">edge_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">lzgraph</span><span class="o">.</span><span class="n">edges</span><span class="p">)</span> 
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;First 10 Nodes: &#39;</span><span class="p">,</span><span class="nb">list</span><span class="p">(</span><span class="n">lzgraph</span><span class="o">.</span><span class="n">nodes</span><span class="p">)[:</span><span class="mi">10</span><span class="p">])</span>
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;First 10 Edges: &#39;</span><span class="p">,</span><span class="nb">list</span><span class="p">(</span><span class="n">lzgraph</span><span class="o">.</span><span class="n">edges</span><span class="p">)[:</span><span class="mi">10</span><span class="p">])</span>
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>First 10 Nodes:  [&#39;C_1&#39;, &#39;A_2&#39;, &#39;S_3&#39;, &#39;SQ_5&#39;, &#39;Q_6&#39;, &#39;G_7&#39;, &#39;R_8&#39;, &#39;D_9&#39;, &#39;T_10&#39;, &#39;QY_12&#39;]
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>First 10 Edges:  [(&#39;C_1&#39;, &#39;A_2&#39;), (&#39;C_1&#39;, &#39;T_2&#39;), (&#39;C_1&#39;, &#39;V_2&#39;), (&#39;C_1&#39;, &#39;S_2&#39;), (&#39;C_1&#39;, &#39;R_2&#39;), (&#39;C_1&#39;, &#39;G_2&#39;), (&#39;C_1&#39;, &#39;D_2&#39;), (&#39;C_1&#39;, &#39;F_2&#39;), (&#39;C_1&#39;, &#39;I_2&#39;), (&#39;A_2&#39;, &#39;S_3&#39;)]
</code></pre></div></p>
<h4 id="distribution-of-observed-sequence-lengths">Distribution of Observed Sequence Lengths:</h4>
<p>After fitting an LZGraph to a repertoire you can quickly access the
distribution of different sequence lengths observed in the repertoire:</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="nb">print</span><span class="p">(</span><span class="n">lzgraph</span><span class="o">.</span><span class="n">lengths</span><span class="p">)</span>
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="p">{</span><span class="mi">13</span><span class="p">:</span> <span class="mi">2973</span><span class="p">,</span> <span class="mi">15</span><span class="p">:</span> <span class="mi">5075</span><span class="p">,</span> <span class="mi">14</span><span class="p">:</span> <span class="mi">4412</span><span class="p">,</span> <span class="mi">16</span><span class="p">:</span> <span class="mi">2862</span><span class="p">,</span>
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a> <span class="mi">12</span><span class="p">:</span> <span class="mi">1147</span><span class="p">,</span> <span class="mi">17</span><span class="p">:</span> <span class="mi">1401</span><span class="p">,</span> <span class="mi">19</span><span class="p">:</span> <span class="mi">268</span><span class="p">,</span> <span class="mi">22</span><span class="p">:</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">20</span><span class="p">:</span> <span class="mi">108</span><span class="p">,</span>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a> <span class="mi">11</span><span class="p">:</span> <span class="mi">508</span><span class="p">,</span> <span class="mi">18</span><span class="p">:</span> <span class="mi">655</span><span class="p">,</span> <span class="mi">21</span><span class="p">:</span> <span class="mi">38</span><span class="p">,</span> <span class="mi">10</span><span class="p">:</span> <span class="mi">72</span><span class="p">,</span>
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a> <span class="mi">24</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">9</span><span class="p">:</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">7</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span><span class="mi">8</span><span class="p">:</span> <span class="mi">1</span><span class="p">}</span>
</code></pre></div>
Note:
The key of the dictionary is an observed sequence length, the value corresponds to the
total number of sequences observed matching the key length.</p>
<h4 id="exploring-graph-initial-and-terminal-states">Exploring Graph Initial and Terminal States:</h4>
<p>Initial state is defined as the first sub-pattern observed in a sequence,
 Terminal state is the last sub-pattern observed in a sequence, the lists and counts
 of all observed initial and final states are saved in every LZGraph and can be 
 accessed:</p>
<h5 id="getting-all-the-initial-states">Getting all the initial states</h5>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="c1"># Getting all the initial states</span>
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a><span class="n">init_states</span> <span class="o">=</span> <span class="n">lzgraph</span><span class="o">.</span><span class="n">initial_states</span>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="nb">print</span><span class="p">(</span><span class="n">init_states</span><span class="p">)</span>
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="n">C_1</span>    <span class="mi">19523</span>
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a><span class="n">R_1</span>       <span class="mi">20</span>
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a><span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>
</code></pre></div></p>
<h5 id="getting-all-the-terminal-states">Getting all the terminal states</h5>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="n">term_states</span> <span class="o">=</span> <span class="n">lzgraph</span><span class="o">.</span><span class="n">print</span><span class="p">(</span><span class="n">lzgraph</span><span class="o">.</span><span class="n">terminal_states</span><span class="p">)</span>
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a><span class="nb">print</span><span class="p">(</span><span class="n">term_states</span><span class="p">)</span>
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="n">F_13</span>      <span class="mi">2672</span>
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a><span class="n">F_15</span>      <span class="mi">4490</span>
<a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a><span class="n">F_14</span>      <span class="mi">3860</span>
<a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a><span class="n">F_16</span>      <span class="mi">2502</span>
<a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a><span class="n">F_12</span>      <span class="mi">1028</span>
<a id="__codelineno-10-6" name="__codelineno-10-6" href="#__codelineno-10-6"></a>          <span class="o">...</span> 
<a id="__codelineno-10-7" name="__codelineno-10-7" href="#__codelineno-10-7"></a><span class="n">G_9</span>          <span class="mi">1</span>
<a id="__codelineno-10-8" name="__codelineno-10-8" href="#__codelineno-10-8"></a><span class="n">AFF_12</span>       <span class="mi">1</span>
<a id="__codelineno-10-9" name="__codelineno-10-9" href="#__codelineno-10-9"></a><span class="n">FF_11</span>        <span class="mi">1</span>
<a id="__codelineno-10-10" name="__codelineno-10-10" href="#__codelineno-10-10"></a><span class="n">F_8</span>          <span class="mi">1</span>
<a id="__codelineno-10-11" name="__codelineno-10-11" href="#__codelineno-10-11"></a><span class="n">HF_22</span>        <span class="mi">1</span>
<a id="__codelineno-10-12" name="__codelineno-10-12" href="#__codelineno-10-12"></a><span class="n">Length</span><span class="p">:</span> <span class="mi">86</span><span class="p">,</span> <span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>
</code></pre></div></p>
<h4 id="graph-v-and-j-marginal-distributions">Graph V and J Marginal Distributions:</h4>
<p>If V and J annotation were provided while constructing the graph,
the genetic data was embedded into each edge of the graph.
The marginal probability of each gene/allele can be observed by the following way:
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="c1"># print(lzgraph.marginal_vgenes)</span>
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a><span class="nb">print</span><span class="p">(</span><span class="n">lzgraph</span><span class="o">.</span><span class="n">marginal_jgenes</span><span class="p">)</span>
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>TRBJ1-2*01    0.159804
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a>TRBJ1-1*01    0.143588
<a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a>TRBJ2-7*01    0.141951
<a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a>TRBJ2-1*01    0.090081
<a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a>TRBJ2-3*01    0.088086
<a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a>TRBJ1-5*01    0.076986
<a id="__codelineno-12-7" name="__codelineno-12-7" href="#__codelineno-12-7"></a>TRBJ2-5*01    0.075963
<a id="__codelineno-12-8" name="__codelineno-12-8" href="#__codelineno-12-8"></a>TRBJ2-2*01    0.057343
<a id="__codelineno-12-9" name="__codelineno-12-9" href="#__codelineno-12-9"></a>TRBJ1-4*01    0.050437
<a id="__codelineno-12-10" name="__codelineno-12-10" href="#__codelineno-12-10"></a>TRBJ1-6*01    0.047675
<a id="__codelineno-12-11" name="__codelineno-12-11" href="#__codelineno-12-11"></a>TRBJ1-3*01    0.043787
<a id="__codelineno-12-12" name="__codelineno-12-12" href="#__codelineno-12-12"></a>TRBJ2-6*01    0.015500
<a id="__codelineno-12-13" name="__codelineno-12-13" href="#__codelineno-12-13"></a>TRBJ2-4*01    0.008798
<a id="__codelineno-12-14" name="__codelineno-12-14" href="#__codelineno-12-14"></a>Name: J, dtype: float64
</code></pre></div></p>
<h2 id="calculating-the-lzcentrality">Calculating the LZCentrality</h2>
<p>The <code>LZGraph</code> library provides a function, <code>LZCentrality</code>, for calculating the LZCentrality of a given CDR3 sequence in a repertoire represented by an LZGraph. The LZCentrality is a measure of sequence centrality within a repertoire, as presented in our paper.</p>
<p>Here's how you can use the <code>LZCentrality</code> function:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="kn">from</span> <span class="nn">LZGraphs</span> <span class="kn">import</span> <span class="n">LZCentrality</span><span class="p">,</span> <span class="n">NDPLZGraph</span>
<a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a>
<a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a><span class="c1"># Assume we have a list of CDR3 sequences and a sequence of interest</span>
<a id="__codelineno-13-4" name="__codelineno-13-4" href="#__codelineno-13-4"></a><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;sequence1&quot;</span><span class="p">,</span> <span class="s2">&quot;sequence2&quot;</span><span class="p">,</span> <span class="s2">&quot;sequence3&quot;</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
<a id="__codelineno-13-5" name="__codelineno-13-5" href="#__codelineno-13-5"></a><span class="n">sequence_of_interest</span> <span class="o">=</span> <span class="s2">&quot;sequence1&quot;</span>
<a id="__codelineno-13-6" name="__codelineno-13-6" href="#__codelineno-13-6"></a>
<a id="__codelineno-13-7" name="__codelineno-13-7" href="#__codelineno-13-7"></a><span class="c1"># Create an LZGraph</span>
<a id="__codelineno-13-8" name="__codelineno-13-8" href="#__codelineno-13-8"></a><span class="n">graph</span> <span class="o">=</span> <span class="n">NDPLZGraph</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
<a id="__codelineno-13-9" name="__codelineno-13-9" href="#__codelineno-13-9"></a>
<a id="__codelineno-13-10" name="__codelineno-13-10" href="#__codelineno-13-10"></a><span class="c1"># Calculate the LZCentrality</span>
<a id="__codelineno-13-11" name="__codelineno-13-11" href="#__codelineno-13-11"></a><span class="n">lzcentrality</span> <span class="o">=</span> <span class="n">LZCentrality</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">sequence_of_interest</span><span class="p">)</span>
<a id="__codelineno-13-12" name="__codelineno-13-12" href="#__codelineno-13-12"></a><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The LZCentrality of </span><span class="si">{</span><span class="n">sequence_of_interest</span><span class="si">}</span><span class="s2"> is </span><span class="si">{</span><span class="n">lzcentrality</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
<p>In this example, <code>NDPLZGraph</code> is the LZGraph class used. You can replace it with any other LZGraph class, such as <code>AAPLZGraph</code>.</p>
<p>This function provides an efficient and scalable way to calculate the LZCentrality, allowing you to gain important insights into the centrality of a sequence within a TCRB repertoire.</p>
<h2 id="calculating-the-k1000-diversity-index">Calculating the K1000 Diversity Index</h2>
<p>The <code>LZGraph</code> library provides a convenient function, <code>K1000_Diversity</code>, for calculating the K1000 Diversity index of a list of CDR3 sequences. This index, as presented in our paper, offers a new way to measure the diversity of a repertoire.</p>
<p>Here's how you can use the <code>K1000_Diversity</code> function:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="kn">from</span> <span class="nn">LZGraphs.Metircs.Metrics</span> <span class="kn">import</span> <span class="n">K1000_Diversity</span>
<a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a><span class="kn">from</span> <span class="nn">LZGraphs</span> <span class="kn">import</span> <span class="n">AAPLZGraph</span>
<a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a>
<a id="__codelineno-14-4" name="__codelineno-14-4" href="#__codelineno-14-4"></a><span class="c1"># Assume we have a list of CDR3 sequences</span>
<a id="__codelineno-14-5" name="__codelineno-14-5" href="#__codelineno-14-5"></a><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;sequence1&quot;</span><span class="p">,</span> <span class="s2">&quot;sequence2&quot;</span><span class="p">,</span> <span class="s2">&quot;sequence3&quot;</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
<a id="__codelineno-14-6" name="__codelineno-14-6" href="#__codelineno-14-6"></a>
<a id="__codelineno-14-7" name="__codelineno-14-7" href="#__codelineno-14-7"></a><span class="c1"># Calculate the K1000 Diversity index</span>
<a id="__codelineno-14-8" name="__codelineno-14-8" href="#__codelineno-14-8"></a><span class="n">k1000</span> <span class="o">=</span> <span class="n">K1000_Diversity</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">AAPLZGraph</span><span class="o">.</span><span class="n">encode_sequence</span><span class="p">,</span> <span class="n">draws</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<a id="__codelineno-14-9" name="__codelineno-14-9" href="#__codelineno-14-9"></a><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The K1000 Diversity index is </span><span class="si">{</span><span class="n">k1000</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
<p>In this example, <code>AAPLZGraph.encode_sequence</code> is the LZGraph encoding function used. You can replace it with any other LZGraph encoding function, such as <code>NDPLZGraph.encode_sequence</code>.</p>
<p>The <code>draws</code> parameter specifies the number of draws for the resampling test. The higher the value, the more realizations will be generated for the K1000 index, and the returned mean value will be a better approximation.</p>
<p>The <code>K1000_Diversity</code> function works by resampling the list of sequences and building LZGraphs based on the provided encoding function. The resampling is performed <code>draws</code> number of times, with each resampled set containing 1000 unique sequences. For each resampled set, the K1000 Diversity index is calculated using the 'nodes' value of the last item in the result dictionary obtained from the NodeEdgeSaturationProbe's resampling test. The average K1000 Diversity index from all the resampling tests is then returned.</p>
<p>This function provides an efficient and scalable way to calculate the K1000 Diversity index, allowing you to gain important insights into the diversity of a TCRB repertoire.</p>
<h2 id="deriving-sequence-generation-probability-pgen">Deriving Sequence Generation Probability (Pgen)</h2>
<p>After an LZGraph is derived for a given repertoire, one can assess the generation
probability (Pgen) of a new sequence with respect to that repertoire (LZGraph).
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a><span class="n">new_sequence</span> <span class="o">=</span> <span class="s2">&quot;CASRGERGDNEQFF&quot;</span>
<a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a><span class="c1"># encode the sequence into relevant graph format</span>
<a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a><span class="n">encoded_sequence</span> <span class="o">=</span> <span class="n">AAPLZGraph</span><span class="o">.</span><span class="n">encode_sequence</span><span class="p">(</span><span class="n">new_sequence</span><span class="p">)</span>
<a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a><span class="c1">#output</span>
<a id="__codelineno-15-5" name="__codelineno-15-5" href="#__codelineno-15-5"></a><span class="c1"># [&#39;C_1&#39;, &#39;A_2&#39;, &#39;S_3&#39;, &#39;R_4&#39;, &#39;G_5&#39;, &#39;E_6&#39;, &#39;RG_8&#39;, &#39;D_9&#39;, &#39;N_10&#39;, &#39;EQ_12&#39;, &#39;F_13&#39;, &#39;F_14&#39;]</span>
<a id="__codelineno-15-6" name="__codelineno-15-6" href="#__codelineno-15-6"></a>
<a id="__codelineno-15-7" name="__codelineno-15-7" href="#__codelineno-15-7"></a><span class="n">pgen</span> <span class="o">=</span> <span class="n">lzgraph</span><span class="o">.</span><span class="n">walk_probability</span><span class="p">(</span><span class="n">encoded_sequence</span><span class="p">)</span>
<a id="__codelineno-15-8" name="__codelineno-15-8" href="#__codelineno-15-8"></a><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">new_sequence</span><span class="si">}</span><span class="s1"> Pgen: </span><span class="si">{</span><span class="n">pgen</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>CASRGERGDNEQFF Pgen: 6.692834748776949e-13
</code></pre></div></p>
<h2 id="generating-new-sequences">Generating New Sequences</h2>
<p>An LZGraph constructed from a source repertoire can generate new sequences following
the statistical properties observed in the source.
Note that that if V and J genes were provided the generation function will be testing
that the random sequence and their sub-pattern all are associated to initially random
V and J genes selected before generating the sequence.
<div class="highlight"><pre><span></span><code><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a><span class="n">generated_walk</span><span class="p">,</span><span class="n">selected_v</span><span class="p">,</span><span class="n">selected_j</span> <span class="o">=</span> <span class="n">lzgraph</span><span class="o">.</span><span class="n">genomic_random_walk</span><span class="p">()</span>
<a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a><span class="c1"># replace genomic_random_walk with random_walk() if annotation was not provided</span>
<a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a><span class="nb">print</span><span class="p">(</span><span class="n">generated_walk</span><span class="p">,</span><span class="n">selected_v</span><span class="p">,</span><span class="n">selected_j</span><span class="p">)</span>
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a>[&#39;C_1&#39;, &#39;S_2&#39;, &#39;A_3&#39;, &#39;T_4&#39;, &#39;G_5&#39;, &#39;GI_7&#39;, &#39;Q_8&#39;, &#39;Y_9&#39;, &#39;QE_11&#39;, &#39;TQ_13&#39;, &#39;YF_15&#39;],
<a id="__codelineno-18-2" name="__codelineno-18-2" href="#__codelineno-18-2"></a>&#39;TRBV29-1*01&#39;,
<a id="__codelineno-18-3" name="__codelineno-18-3" href="#__codelineno-18-3"></a>&#39;TRBJ2-5*01&#39;
</code></pre></div>
You can transform the resulting walk back into a sequence by cleaning each node and
concatenating them into a single string.
Each LZGraph has a "clean_node" method that accompanies the relevant "encode_sequence"
method.
<div class="highlight"><pre><span></span><code><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a><span class="n">clean_sequence</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">AAPLZGraph</span><span class="o">.</span><span class="n">clean_node</span><span class="p">(</span><span class="n">node</span><span class="p">)</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">generated_walk</span><span class="p">])</span>
<a id="__codelineno-19-2" name="__codelineno-19-2" href="#__codelineno-19-2"></a><span class="c1"># Note that AAPLZGraph needs to be replaced in case you used the NDPLZGraph</span>
<a id="__codelineno-19-3" name="__codelineno-19-3" href="#__codelineno-19-3"></a><span class="nb">print</span><span class="p">(</span><span class="n">clean_sequence</span><span class="p">)</span>
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>CASSPGTGGTGELFF
</code></pre></div></p>
<h2 id="feature-vector-from-graph-eigen-centrality">Feature Vector From Graph (Eigen Centrality)</h2>
<p>One can derive a vector representation of an LZGraph in order to perform any
modeling task. The resulting vector in a sense capture the structural dynamics of 
a repertoire.
Note that the number of features in the resulting vector equal the number of nodes
in the graph, and the value in each node correspond to each node, so it is recommended
using the Naive LZGraph, this allows to create LZGraphs to different repertoires while
maintaining the same dictionary of sub-patterns (nodes).</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a><span class="kn">from</span> <span class="nn">LZGraphs.Graphs.Naive</span> <span class="kn">import</span> <span class="n">NaiveLZGraph</span><span class="p">,</span> <span class="n">generate_dictionary</span>
<a id="__codelineno-21-2" name="__codelineno-21-2" href="#__codelineno-21-2"></a>
<a id="__codelineno-21-3" name="__codelineno-21-3" href="#__codelineno-21-3"></a><span class="c1"># Create a dictionary that will be shared between all repertoires</span>
<a id="__codelineno-21-4" name="__codelineno-21-4" href="#__codelineno-21-4"></a><span class="n">dictionary</span> <span class="o">=</span> <span class="n">generate_dictionary</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span>
<a id="__codelineno-21-5" name="__codelineno-21-5" href="#__codelineno-21-5"></a><span class="c1"># create a graph, this can be done in a loop over many repertoires</span>
<a id="__codelineno-21-6" name="__codelineno-21-6" href="#__codelineno-21-6"></a><span class="n">lzgraph</span> <span class="o">=</span> <span class="n">NaiveLZGraph</span><span class="p">(</span><span class="n">list_of_sequences</span><span class="p">,</span> <span class="n">dictionary</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-21-7" name="__codelineno-21-7" href="#__codelineno-21-7"></a><span class="c1"># derive the feature vector</span>
<a id="__codelineno-21-8" name="__codelineno-21-8" href="#__codelineno-21-8"></a><span class="n">feature_vector</span> <span class="o">=</span> <span class="n">lzgraph</span><span class="o">.</span><span class="n">eigenvector_centrality</span><span class="p">()</span>
<a id="__codelineno-21-9" name="__codelineno-21-9" href="#__codelineno-21-9"></a><span class="nb">print</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">feature_vector</span><span class="p">))</span>
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a>A         3.009520e-01
<a id="__codelineno-22-2" name="__codelineno-22-2" href="#__codelineno-22-2"></a>T         1.183398e-01
<a id="__codelineno-22-3" name="__codelineno-22-3" href="#__codelineno-22-3"></a>G         1.186366e-01
<a id="__codelineno-22-4" name="__codelineno-22-4" href="#__codelineno-22-4"></a>C         2.461758e-01
<a id="__codelineno-22-5" name="__codelineno-22-5" href="#__codelineno-22-5"></a>AA        1.252643e-01
<a id="__codelineno-22-6" name="__codelineno-22-6" href="#__codelineno-22-6"></a>              ...     
<a id="__codelineno-22-7" name="__codelineno-22-7" href="#__codelineno-22-7"></a>CCCCGC    5.252576e-10
<a id="__codelineno-22-8" name="__codelineno-22-8" href="#__codelineno-22-8"></a>CCCCCA    5.252576e-10
<a id="__codelineno-22-9" name="__codelineno-22-9" href="#__codelineno-22-9"></a>CCCCCT    5.252576e-10
<a id="__codelineno-22-10" name="__codelineno-22-10" href="#__codelineno-22-10"></a>CCCCCG    5.252576e-10
<a id="__codelineno-22-11" name="__codelineno-22-11" href="#__codelineno-22-11"></a>CCCCCC    5.252576e-10
<a id="__codelineno-22-12" name="__codelineno-22-12" href="#__codelineno-22-12"></a>Length: 5460, dtype: float64
</code></pre></div></p>
<p>Note:
The <code>generate_dictionary</code> function given the value 6 return <code>5460</code> sub-patterns,
many of which may not exist in your graph, nonetheless this promises that all
the repertoires encoded will produce the same feature vector dimension but the
value across the different sub-pattern will be different.</p>
<h2 id="plotting-graph-related-features">Plotting Graph Related Features</h2>
<p>The library provides multiple plots that can be useful when analyzing sequence
and comparing between sequences.</p>
<h3 id="ancestors-descendants-curves-plot">Ancestors Descendants Curves Plot</h3>
<p>In this chart you can examine at each sub-pattern of a given sequence
the amount descendants nodes reachable at each node contained in your sequence,
as well as the amount of ancestors at each node.
Sequence with different attribute differ both by the slope and the convergence
rate of these curve as well as by the intersection point between the curves.</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a><span class="kn">from</span> <span class="nn">LZGraphs.Visualization.Visualize</span> <span class="kn">import</span> <span class="n">ancestors_descendants_curves_plot</span>
<a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a>
<a id="__codelineno-23-3" name="__codelineno-23-3" href="#__codelineno-23-3"></a><span class="n">sequence</span> <span class="o">=</span> <span class="s1">&#39;CASTPGTASGYTF&#39;</span>
<a id="__codelineno-23-4" name="__codelineno-23-4" href="#__codelineno-23-4"></a><span class="n">ancestors_descendants_curves_plot</span><span class="p">(</span><span class="n">lzgraph</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</code></pre></div>
Output:
<img alt="alt text" src="../images/ad_curve_example.png" /></p>
<h3 id="sequence-possible-paths-plot">Sequence Possible Paths Plot</h3>
<p>In this chart we look at a reduced and immediate version of Descendants curve.
For each sub-pattern derived from a given sequence and based on an LZGraph,
We can examine the number of alternatives there are at each node, this indicates
the rarity of a sequence and is correlated with the difference from the 
mean Levenshtein distance of the repertoire as shown in the LZGraphs paper.</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a><span class="kn">from</span> <span class="nn">LZGraphs.Visualization.Visualize</span> <span class="kn">import</span> <span class="n">sequence_possible_paths_plot</span>
<a id="__codelineno-24-2" name="__codelineno-24-2" href="#__codelineno-24-2"></a>
<a id="__codelineno-24-3" name="__codelineno-24-3" href="#__codelineno-24-3"></a><span class="n">sequence</span> <span class="o">=</span> <span class="s1">&#39;CASTPGTASGYTF&#39;</span>
<a id="__codelineno-24-4" name="__codelineno-24-4" href="#__codelineno-24-4"></a><span class="n">sequence_possible_paths_plot</span><span class="p">(</span><span class="n">lzgraph</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</code></pre></div>
Output:
<img alt="alt text" src="../images/sequence_path_number_example.png" /></p>
<h3 id="node-genomic-variability-plot">Node Genomic Variability Plot</h3>
<p>In this chart we look at the number of V and J genes/alleles per node in a given sequence with respect
to a given repertoire.
Not only can one infer the sub-patterns in a sequence that have the exceptional number of V and J alternatives
but also when comparing between the same sequence in different repertoires (different LZGraphs) one can infer
the amount difference at each sub-pattern between the two repertoires.</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a><span class="kn">from</span> <span class="nn">LZGraphs.Visualization.Visualize</span> <span class="kn">import</span> <span class="n">sequence_genomic_node_variability_plot</span>
<a id="__codelineno-25-2" name="__codelineno-25-2" href="#__codelineno-25-2"></a>
<a id="__codelineno-25-3" name="__codelineno-25-3" href="#__codelineno-25-3"></a><span class="n">sequence</span> <span class="o">=</span> <span class="s1">&#39;CASTPGTASGYTF&#39;</span>
<a id="__codelineno-25-4" name="__codelineno-25-4" href="#__codelineno-25-4"></a><span class="n">sequence_genomic_node_variability_plot</span><span class="p">(</span><span class="n">lzgraph</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</code></pre></div>
Output:
<img alt="alt text" src="../images/number_of_vj_at_nodes_example.png" /></p>
<h3 id="edge-genomic-variability-plot">Edge Genomic Variability Plot</h3>
<p>In this chart we look at the number of V and J genes/alleles per edge in a given sequence with respect
to a given repertoire.</p>
<ul>
<li>Allele/gene names colored in red signify that the allele/gene appeared in all the edges in the given sequence.</li>
<li>Black cells signify that this spesific allele/gene wasnt observed at that edge.</li>
<li>The color gradient at each cell represents the probability of choosing that edge under the constraint of having
 that specific V/J.</li>
</ul>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a><span class="kn">from</span> <span class="nn">src.LZGraphs.Visualization.Visualize</span> <span class="kn">import</span> <span class="n">sequence_genomic_edges_variability_plot</span>
<a id="__codelineno-26-2" name="__codelineno-26-2" href="#__codelineno-26-2"></a>
<a id="__codelineno-26-3" name="__codelineno-26-3" href="#__codelineno-26-3"></a><span class="n">sequence</span> <span class="o">=</span> <span class="s1">&#39;CASTPGTASGYTF&#39;</span>
<a id="__codelineno-26-4" name="__codelineno-26-4" href="#__codelineno-26-4"></a><span class="n">sequence_genomic_edges_variability_plot</span><span class="p">(</span><span class="n">lzgraph</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</code></pre></div>
Output:
<img alt="alt text" src="../images/number_of_vj_at_edges_example.png" /></p>
<h2 id="bow-vectorizer">BOW Vectorizer</h2>
<p>The library also provides a convenient class implementing the BOW logic presented in the paper.
Note that each instance of the BOW wrapper class has to fitted on a list of sequences (repertoire) before it
can transform new lists of sequences into BOW vectors.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a><span class="kn">from</span> <span class="nn">src.LZGraphs</span> <span class="kn">import</span> <span class="n">LZBOW</span>
<a id="__codelineno-27-2" name="__codelineno-27-2" href="#__codelineno-27-2"></a><span class="kn">from</span> <span class="nn">src.LZGraphs</span> <span class="kn">import</span> <span class="n">NDPLZGraph</span>
<a id="__codelineno-27-3" name="__codelineno-27-3" href="#__codelineno-27-3"></a>
<a id="__codelineno-27-4" name="__codelineno-27-4" href="#__codelineno-27-4"></a><span class="n">sequence_list</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">cdr3_rearrangement</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
<a id="__codelineno-27-5" name="__codelineno-27-5" href="#__codelineno-27-5"></a>
<a id="__codelineno-27-6" name="__codelineno-27-6" href="#__codelineno-27-6"></a><span class="c1"># create vectorizer and choose the Nucleotide Double Positional (ndp) encdoing function (default is Naive)</span>
<a id="__codelineno-27-7" name="__codelineno-27-7" href="#__codelineno-27-7"></a><span class="n">vectorizer</span> <span class="o">=</span> <span class="n">LZBOW</span><span class="p">(</span><span class="n">encoding_function</span><span class="o">=</span><span class="n">NDPLZGraph</span><span class="o">.</span><span class="n">encode_sequence</span><span class="p">)</span>
<a id="__codelineno-27-8" name="__codelineno-27-8" href="#__codelineno-27-8"></a>
<a id="__codelineno-27-9" name="__codelineno-27-9" href="#__codelineno-27-9"></a><span class="c1"># fit on sequence list</span>
<a id="__codelineno-27-10" name="__codelineno-27-10" href="#__codelineno-27-10"></a><span class="n">vectorizer</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">sequence_list</span><span class="p">)</span>
<a id="__codelineno-27-11" name="__codelineno-27-11" href="#__codelineno-27-11"></a>
<a id="__codelineno-27-12" name="__codelineno-27-12" href="#__codelineno-27-12"></a><span class="c1"># BOW dictionary</span>
<a id="__codelineno-27-13" name="__codelineno-27-13" href="#__codelineno-27-13"></a><span class="n">vectorizer</span><span class="o">.</span><span class="n">dictionary</span>
<a id="__codelineno-27-14" name="__codelineno-27-14" href="#__codelineno-27-14"></a>
<a id="__codelineno-27-15" name="__codelineno-27-15" href="#__codelineno-27-15"></a><span class="n">bow_vector</span> <span class="o">=</span> <span class="n">vectorizer</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">new_list_of_sequences</span><span class="p">)</span>
</code></pre></div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.bd41221c.min.js"></script>
      
    
  </body>
</html>